client/api_object.go

```diff
@@ -353,6 +359,279 @@ func (c *client) FGetObject(ctx context.Context, bucketName, objectName, filePat
 	return nil
 }
 
+func (c *client) RecoverObjectBySecondary(ctx context.Context, bucketName, objectName, filePath string, opts types.GetObjectOption) error {
+	// compute segment count
+	dataBlocks, parityBlocks, maxSegmentSize, err := c.GetRedundancyParams()
+	if err != nil {
+		return err
+	}
+	ecPieceCount := dataBlocks + parityBlocks
+	minRecoveryPieces := dataBlocks
+
+	objectInfo, err := c.HeadObject(ctx, bucketName, objectName)
+	if err != nil {
+		return err
+	}
+	objectSize := objectInfo.PayloadSize
+	startSegmentIdx, endSegmentIdx, diffStartOffset, diffEndOffset, err := checkGetObjectRange(objectSize, maxSegmentSize, opts)
+	if err != nil {
+		return err
+	}
+
+	secondaryEndpoints, err := c.getSecondaryEndpoints(ctx, objectInfo)
+	if err != nil {
+		return err
+	}
+
+	_, err = os.Stat(filePath)
+	if err == nil {
+		return errors.New("download file already exist:" + filePath)
+	}
+
+	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// iterate over segment indices and recover one by one
+	for segmentIdx := startSegmentIdx; segmentIdx <= endSegmentIdx; segmentIdx++ {
+		doneCh := make(chan bool, len(secondaryEndpoints))
+		quitCh := make(chan bool)
+		totalTaskNum := int32(ecPieceCount)
+		doneTaskNum := uint32(0)
+		downLoadPieceSize := 0
+		var recoveryDataSources = make([][]byte, ecPieceCount)
+
+		segmentSize := utils.GetSegmentSize(objectSize, uint32(segmentIdx), maxSegmentSize)
+		for ecIdx := 0; ecIdx < int(ecPieceCount); ecIdx++ {
+			recoveryDataSources[ecIdx] = nil
+			go func(secondaryIndex int) {
```

Comment:
Add TODO to compare.  For now,  since the network delay overhead is far greater than the decode overhead , priority is given here to the secondary with low delay.  The  16M segment decode with parity chunks  cost 1ms  which is not so different with data chunks